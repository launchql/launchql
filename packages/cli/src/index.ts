#!/usr/bin/env node
import { CLI, CLIOptions } from 'inquirerer';

import { commands } from './commands';

export const options: Partial<CLIOptions> = {
  minimistOpts: {
    alias: {
      v: 'version',
      h: 'help'
    }
  }
};

const app = new CLI(commands, options);

app.run().then(()=> {
  // all done!
}).catch(error => {
  // Should not reach here with the new CLI error handling pattern
  // But keep as fallback for unexpected errors
  console.error('Unexpected error:', error);
  process.exit(1);
});