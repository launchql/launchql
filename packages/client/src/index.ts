import { Pool, PoolClient } from 'pg';

import env from './env';

export class Database {
  private static instance: Database;
  private pool: Pool;

  constructor() {
    if (Database.instance) {
      return Database.instance;
    }

    const pgPoolConfig = {
      connectionString: env.DATABASE_URL,
    };
    this.pool = new Pool(pgPoolConfig);

    // Ensure the pool is closed on process termination
    process.on('SIGTERM', async () => {
      await this.shutdown();
    });

    Database.instance = this;
    return this;
  }

  /**
   * Executes a callback function within a database transaction.
   * @param fn - A callback function that receives a PoolClient to perform database operations.
   */
  async withTransaction(fn: (client: PoolClient) => Promise<void>): Promise<void> {
    const client = await this.pool.connect();
    try {
      await client.query('BEGIN');
      try {
        await fn(client);
        await client.query('COMMIT');
      } catch (e) {
        console.error('Error during transaction:', e);
        await client.query('ROLLBACK');
        throw e; // Re-throw the error to propagate it
      }
    } finally {
      client.release();
    }
  }

  /**
   * Shuts down the connection pool.
   */
  async shutdown(): Promise<void> {
    await this.pool.end();
  }
}