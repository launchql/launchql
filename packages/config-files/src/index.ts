export * from './types';
export * from './plan';
export * from './config';
export * from './sql';
export * from './extension';
export * from './sql-scripts';