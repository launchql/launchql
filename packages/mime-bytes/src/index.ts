// Main export file for mime-bytes package

// Export the main detector class and convenience functions
export {
  FileTypeDetector,
  FileTypeDetectorOptions,
  defaultDetector,
  detectFromStream,
  detectFromBuffer,
  detectFromExtension
} from './file-type-detector';

// Export registry types and functions
export {
  FileTypeDefinition,
  DetectionResult,
  ContentTypeMapping,
  FILE_TYPES,
  CONTENT_TYPE_MAPPINGS,
  getFileTypeByMagicBytes,
  getFileTypeByExtension,
  getFileTypesByCategory,
  getContentTypeByExtension,
  detectCharset
} from './file-types-registry';

// Export peek stream functionality
export {
  peek,
  BufferPeekStream,
  PeekStreamOptions,
  PeekCallback,
  PeekPromise
} from './peak';

// Export utility functions
export * from './utils/magic-bytes';
export * from './utils/mime-types';
export * from './utils/extensions';