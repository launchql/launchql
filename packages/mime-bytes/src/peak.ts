// TypeScript implementation of peek stream for efficient file type detection
// Based on reference-packages/buffer-peak/peak.js

import { Transform, TransformOptions, Readable } from 'stream';
import { EventEmitter } from 'events';

export interface PeekStreamOptions extends TransformOptions {
  peekBytes: number;
}

export class BufferPeekStream extends Transform {
  private peekBytes: number;
  private buffer: Buffer;
  private bufferLength: number;
  private peeked: boolean;

  constructor(options: PeekStreamOptions) {
    super(options);
    this.peekBytes = options.peekBytes || 16;
    this.buffer = Buffer.alloc(0);
    this.bufferLength = 0;
    this.peeked = false;
  }

  _transform(chunk: any, encoding: BufferEncoding, callback: Function): void {
    if (this.peeked) {
      // After peeking, just pass through
      this.push(chunk);
      callback();
      return;
    }

    // Accumulate data until we have enough to peek
    const chunkBuffer = Buffer.isBuffer(chunk) ? chunk : Buffer.from(chunk, encoding);
    this.buffer = Buffer.concat([this.buffer, chunkBuffer]);
    this.bufferLength += chunkBuffer.length;

    if (this.bufferLength >= this.peekBytes) {
      // We have enough data to peek
      this.peeked = true;
      
      // Emit the peek event with the requested bytes
      const peekBuffer = this.buffer.slice(0, this.peekBytes);
      this.emit('peek', peekBuffer);
      
      // Push all accumulated data
      this.push(this.buffer);
      this.buffer = Buffer.alloc(0);
      this.bufferLength = 0;
      
      callback();
    } else {
      // Need more data
      callback();
    }
  }

  _flush(callback: Function): void {
    if (!this.peeked && this.bufferLength > 0) {
      // Not enough data was received, emit what we have
      this.peeked = true;
      this.emit('peek', this.buffer);
      this.push(this.buffer);
    }
    callback();
  }
}

// Main peek function with callback
export function peek(source: Readable, bytes: number, callback?: (err: Error | null, buffer: Buffer, dest: BufferPeekStream) => void): BufferPeekStream;
export function peek(source: Readable, callback: (err: Error | null, buffer: Buffer, dest: BufferPeekStream) => void): BufferPeekStream;
export function peek(
  source: Readable,
  bytesOrCallback: number | ((err: Error | null, buffer: Buffer, dest: BufferPeekStream) => void),
  callback?: (err: Error | null, buffer: Buffer, dest: BufferPeekStream) => void
): BufferPeekStream {
  let bytes: number;
  let cb: ((err: Error | null, buffer: Buffer, dest: BufferPeekStream) => void) | undefined;

  if (typeof bytesOrCallback === 'function') {
    bytes = 16; // Default peek bytes
    cb = bytesOrCallback;
  } else {
    bytes = bytesOrCallback;
    cb = callback;
  }

  const dest = new BufferPeekStream({ peekBytes: bytes });

  if (cb) {
    dest.once('peek', (buffer: Buffer) => {
      cb(null, buffer, dest);
    });

    dest.once('error', (err: Error) => {
      cb(err, Buffer.alloc(0), dest);
    });
  }

  return source.pipe(dest);
}

// Promise-based version
peek.promise = function(source: Readable, bytes: number = 16): Promise<[Buffer, BufferPeekStream]> {
  return new Promise((resolve, reject) => {
    const dest = peek(source, bytes, (err, buffer, stream) => {
      if (err) {
        reject(err);
      } else {
        resolve([buffer, stream]);
      }
    });

    // Handle source errors
    source.once('error', reject);
  });
};

// Export types
export type PeekCallback = (err: Error | null, buffer: Buffer, dest: BufferPeekStream) => void;
export type PeekPromise = (source: Readable, bytes?: number) => Promise<[Buffer, BufferPeekStream]>;