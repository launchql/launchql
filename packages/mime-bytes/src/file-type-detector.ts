// Main file type detector class with stream-focused API

import { Readable } from 'stream';

import {
  detectCharset,
  DetectionResult,
  FILE_TYPES,
  FileTypeDefinition,
  getContentTypeByExtension,
  getContentTypeForExtension,
  getFileTypeByExtension,
  getFileTypesByCategory} from './file-types-registry';
import {peek } from './peak';
import {normalizeExtension } from './utils/extensions';
import { compareBytes } from './utils/magic-bytes';
import { resolveMimeAlias } from './utils/mime-types';

export interface FileTypeDetectorOptions {
  peekBytes?: number; // Number of bytes to peek (default: 32)
  checkMultipleOffsets?: boolean; // Check multiple offsets (default: true)
  maxOffset?: number; // Maximum offset to check (default: 12)
}

export class FileTypeDetector {
  private fileTypes: FileTypeDefinition[];
  private options: Required<FileTypeDetectorOptions>;
  private magicBytesCache: Map<string, FileTypeDefinition | null>;
  private extensionCache: Map<string, FileTypeDefinition[]>;

  constructor(options: FileTypeDetectorOptions = {}) {
    // Create a copy of FILE_TYPES to avoid modifying the global registry
    this.fileTypes = [...FILE_TYPES];
    this.options = {
      peekBytes: options.peekBytes || 32,
      checkMultipleOffsets: options.checkMultipleOffsets !== false,
      maxOffset: options.maxOffset || 12
    };
    this.magicBytesCache = new Map();
    this.extensionCache = new Map();
  }

  /**
   * Detect file type from a stream (PRIMARY METHOD - memory efficient)
   * @param stream - Readable stream to detect from
   * @returns Detection result or null if not detected
   */
  async detectFromStream(stream: Readable): Promise<DetectionResult | null> {
    try {
      const [buffer, peekStream] = await peek.promise(stream, this.options.peekBytes);
      
      // Check multiple offsets for different file types
      const result = await this.detectFromBuffer(buffer);
      
      // Important: Return the peek stream so it can be used for further processing
      // The caller should use peekStream instead of the original stream
      if (result) {
        (result as any)._stream = peekStream;
      }
      
      return result;
    } catch (error) {
      // Handle stream errors gracefully
      // Only log errors in non-test environments
      if (process.env.NODE_ENV !== 'test') {
        console.error('Error detecting file type from stream:', error);
      }
      return null;
    }
  }

  /**
   * Detect file type from an already-read buffer
   * @param buffer - Buffer to detect from
   * @returns Detection result or null if not detected
   */
  async detectFromBuffer(buffer: Buffer): Promise<DetectionResult | null> {
    if (!buffer || buffer.length === 0) {
      return null;
    }

    // Check multiple offsets if enabled
    const offsets = this.options.checkMultipleOffsets 
      ? this.generateOffsets(buffer.length)
      : [0];

    for (const offset of offsets) {
      const fileType = this.checkMagicBytesAtOffset(buffer, offset);
      if (fileType) {
        return this.enhanceDetectionResult(fileType, buffer);
      }
    }

    // No magic bytes matched, but we can still detect charset for unknown files
    const charset = detectCharset(buffer);
    if (charset !== 'binary') {
      // Return a generic text file result
      return {
        name: 'text',
        mimeType: 'text/plain',
        extensions: ['txt'],
        charset,
        contentType: 'text/plain',
        confidence: 0.5 // Lower confidence since we only detected charset
      };
    }

    return null;
  }

  /**
   * Detect file type from extension only
   * @param extension - File extension (with or without dot)
   * @returns Array of possible detection results with lower confidence
   */
  detectFromExtension(extension: string): DetectionResult[] {
    const cleanExt = normalizeExtension(extension);
    
    // Check cache first
    if (this.extensionCache.has(cleanExt)) {
      const cachedTypes = this.extensionCache.get(cleanExt)!;
      return cachedTypes.map(fileType => ({
        name: fileType.name,
        mimeType: resolveMimeAlias(fileType.mimeType),
        extensions: fileType.extensions,
        charset: 'unknown', // Can't determine charset from extension alone
        contentType: getContentTypeByExtension(cleanExt) || fileType.mimeType,
        confidence: 0.8 // Lower confidence for extension-only detection
      }));
    }
    
    const fileTypes = getFileTypeByExtension(cleanExt);
    
    // Cache the result
    this.extensionCache.set(cleanExt, fileTypes);
    
    return fileTypes.map(fileType => ({
      name: fileType.name,
      mimeType: resolveMimeAlias(fileType.mimeType),
      extensions: fileType.extensions,
      charset: 'unknown', // Can't determine charset from extension alone
      contentType: getContentTypeByExtension(cleanExt) || fileType.mimeType,
      confidence: 0.8 // Lower confidence for extension-only detection
    }));
  }

  /**
   * Get all file types by category
   * @param category - Category name (e.g., 'image', 'video', 'archive')
   * @returns Array of file type definitions
   */
  getByCategory(category: string): FileTypeDefinition[] {
    return getFileTypesByCategory(category);
  }

  /**
   * Add a new file type dynamically
   * @param fileType - File type definition to add
   */
  addFileType(fileType: FileTypeDefinition): void {
    this.fileTypes.push(fileType);
    // Clear caches when file types change
    this.clearCache();
  }

  /**
   * Remove a file type by name
   * @param name - Name of the file type to remove
   */
  removeFileType(name: string): boolean {
    const index = this.fileTypes.findIndex(ft => ft.name === name);
    if (index !== -1) {
      this.fileTypes.splice(index, 1);
      // Clear caches when file types change
      this.clearCache();
      return true;
    }
    return false;
  }

  /**
   * Clear all caches
   */
  clearCache(): void {
    this.magicBytesCache.clear();
    this.extensionCache.clear();
  }

  /**
   * Get all registered file types
   * @returns Array of all file type definitions
   */
  getAllFileTypes(): FileTypeDefinition[] {
    return [...this.fileTypes];
  }

  /**
   * Check magic bytes at a specific offset
   * @private
   */
  private checkMagicBytesAtOffset(buffer: Buffer, offset: number): FileTypeDefinition | null {
    // Check each file type in order
    for (const fileType of this.fileTypes) {
      // Skip if this file type requires a different offset
      if (fileType.offset !== undefined && fileType.offset !== offset) continue;
      
      // Skip if offset is not 0 and file type doesn't specify an offset
      if (offset > 0 && fileType.offset === undefined) continue;
      
      // Check if magic bytes match
      if (compareBytes(buffer, fileType.magicBytes, offset)) {
        return fileType;
      }
    }
    
    return null;
  }

  /**
   * Generate offsets to check based on buffer size
   * @private
   */
  private generateOffsets(bufferLength: number): number[] {
    const offsets: number[] = [];
    for (let i = 0; i <= this.options.maxOffset && i < bufferLength; i += 4) {
      offsets.push(i);
    }
    return offsets;
  }

  /**
   * Enhance detection result with additional information
   * @private
   */
  private enhanceDetectionResult(fileType: FileTypeDefinition, buffer: Buffer): DetectionResult {
    // Use charset from file type definition if available, otherwise detect it
    const charset = fileType.charset || detectCharset(buffer);
    
    // Determine content type based on extension and charset
    let contentType = fileType.contentType || fileType.mimeType;
    
    // Try charset-aware content type lookup first
    if (fileType.extensions.length > 0) {
      const primaryExt = fileType.extensions[0];
      const charsetAwareContentType = getContentTypeForExtension(primaryExt, charset);
      
      if (charsetAwareContentType) {
        contentType = charsetAwareContentType;
      } else if (!fileType.contentType) {
        // Fall back to regular content type lookup if no charset-specific match
        const inferredContentType = getContentTypeByExtension(primaryExt);
        if (inferredContentType) {
          contentType = inferredContentType;
        }
      }
    }

    return {
      name: fileType.name,
      mimeType: resolveMimeAlias(fileType.mimeType),
      extensions: fileType.extensions,
      charset,
      contentType: resolveMimeAlias(contentType),
      confidence: 1.0 // High confidence for magic bytes detection
    };
  }

  /**
   * Detect file type with fallback to extension
   * @param input - Readable stream or Buffer
   * @param filename - Optional filename for extension fallback
   * @returns Detection result with attached stream for reuse (if input was stream)
   */
  async detectWithFallback(input: Readable | Buffer, filename?: string): Promise<(DetectionResult & { _stream?: Readable }) | null> {
    try {
      let buffer: Buffer;
      let peekStream: Readable | undefined;
      
      // Handle both Buffer and Readable inputs
      if (Buffer.isBuffer(input)) {
        buffer = input;
      } else {
        const peekResult = await peek.promise(input, this.options.peekBytes);
        buffer = peekResult[0];
        peekStream = peekResult[1];
      }
      
      // Try magic bytes detection first
      const magicResult = await this.detectFromBuffer(buffer);
      
      if (magicResult) {
        // If we have a filename, try to enhance with more specific content type
        if (filename) {
          const lastDot = filename.lastIndexOf('.');
          if (lastDot !== -1) {
            const extension = filename.substring(lastDot + 1);
            
            // Check for generic text files that might have specific content types
            if (magicResult.mimeType === 'text/plain' && magicResult.charset) {
              const contentType = getContentTypeForExtension(extension, magicResult.charset);
              if (contentType) {
                // Enhance the result with charset-aware content type
                const enhancedResult = {
                  ...magicResult,
                  contentType,
                  confidence: 0.8 // Higher confidence since we have both magic bytes and extension
                };
                return peekStream ? { ...enhancedResult, _stream: peekStream } : enhancedResult;
              }
            }
            
            // Check for ZIP files that might be Office Open XML or other specific formats
            if (magicResult.name === 'zip' || magicResult.mimeType === 'application/zip') {
              const contentType = getContentTypeForExtension(extension, magicResult.charset || 'binary');
              if (contentType && contentType !== 'application/zip') {
                // Enhance the result with more specific content type
                const enhancedResult = {
                  ...magicResult,
                  contentType,
                  confidence: 0.9 // High confidence for known ZIP-based formats
                };
                return peekStream ? { ...enhancedResult, _stream: peekStream } : enhancedResult;
              }
            }
          }
        }
        
        // Attach the peek stream for reuse if available
        return peekStream ? { ...magicResult, _stream: peekStream } : magicResult;
      }

      // Fallback to extension if filename provided
      if (filename) {
        const lastDot = filename.lastIndexOf('.');
        if (lastDot !== -1) {
          const extension = filename.substring(lastDot + 1);
          const charset = detectCharset(buffer);
          
          // Try charset-aware content type lookup
          const contentType = getContentTypeForExtension(extension, charset);
          
          if (contentType) {
            // Create a result based on extension and detected charset
            const result: DetectionResult = {
              name: extension.toLowerCase(),
              mimeType: contentType,
              extensions: [extension.toLowerCase()],
              charset,
              contentType,
              confidence: 0.7 // Higher confidence when charset matches
            };
            
            return peekStream ? { ...result, _stream: peekStream } : result;
          }
          
          // Fall back to regular extension detection
          const extensionResults = this.detectFromExtension(extension);
          if (extensionResults.length > 0) {
            const result = {
              ...extensionResults[0],
              charset, // Use detected charset
              confidence: 0.6 // Lower confidence for fallback
            };
            
            // Update content type if charset-specific mapping exists
            const charsetContentType = getContentTypeForExtension(extension, charset);
            if (charsetContentType) {
              result.contentType = charsetContentType;
            }
            
            return peekStream ? { ...result, _stream: peekStream } : result;
          }
        }
      }

      // No detection possible
      return null;
    } catch (error) {
      // Only log errors in non-test environments
      if (process.env.NODE_ENV !== 'test') {
        console.error('Error in detectWithFallback:', error);
      }
      return null;
    }
  }

  /**
   * Check if a buffer matches a specific file type
   * @param buffer - Buffer to check
   * @param fileTypeName - Name of the file type to check against
   * @returns True if matches, false otherwise
   */
  isFileType(buffer: Buffer, fileTypeName: string): boolean {
    const fileType = this.fileTypes.find(ft => ft.name === fileTypeName);
    if (!fileType) return false;

    const offset = fileType.offset || 0;
    return compareBytes(buffer, fileType.magicBytes, offset);
  }

  /**
   * Get statistics about registered file types
   * @returns Statistics object
   */
  getStatistics(): {
    totalTypes: number;
    byCategory: Record<string, number>;
    byMimePrefix: Record<string, number>;
    } {
    const stats = {
      totalTypes: this.fileTypes.length,
      byCategory: {} as Record<string, number>,
      byMimePrefix: {} as Record<string, number>
    };

    for (const fileType of this.fileTypes) {
      // Count by category
      const category = fileType.category || 'other';
      stats.byCategory[category] = (stats.byCategory[category] || 0) + 1;

      // Count by MIME prefix
      const mimePrefix = fileType.mimeType.split('/')[0];
      stats.byMimePrefix[mimePrefix] = (stats.byMimePrefix[mimePrefix] || 0) + 1;
    }

    return stats;
  }
}

// Export a default instance for convenience
export const defaultDetector = new FileTypeDetector();

// Convenience functions using the default detector
export async function detectFromStream(stream: Readable): Promise<DetectionResult | null> {
  return defaultDetector.detectFromStream(stream);
}

export async function detectFromBuffer(buffer: Buffer): Promise<DetectionResult | null> {
  return defaultDetector.detectFromBuffer(buffer);
}

export function detectFromExtension(extension: string): DetectionResult[] {
  return defaultDetector.detectFromExtension(extension);
}