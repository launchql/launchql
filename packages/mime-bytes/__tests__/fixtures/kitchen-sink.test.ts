import { createReadStream } from 'fs';
import { glob } from 'glob';
import path from 'path';

import { FileTypeDetector } from '../../src/file-type-detector';

describe('Kitchen Sink Fixtures', () => {
  let detector: FileTypeDetector;
  
  beforeAll(() => {
    detector = new FileTypeDetector();
  });

  test('should detect all kitchen sink files correctly', async () => {
    const fixturePath = path.join(__dirname, '../../../../__fixtures__/kitchen-sink');
    const files = await glob('**/*', { cwd: fixturePath, nodir: true });
    
    const detectionResults: Record<string, any> = {};
    
    for (const file of files) {
      const filePath = path.join(fixturePath, file);
      const stream = createReadStream(filePath);
      
      try {
        const detectedType = await detector.detectWithFallback(stream, file);
        
        if (detectedType) {
          detectionResults[file] = {
            contentType: detectedType.contentType,
            magic: {
              charset: detectedType.charset,
              type: detectedType.mimeType,
              name: detectedType.name,
              confidence: detectedType.confidence
            }
          };
        } else {
          detectionResults[file] = {
            contentType: null,
            magic: null
          };
        }
      } catch (error) {
        detectionResults[file] = {
          contentType: null,
          magic: null,
          error: error instanceof Error ? error.message : 'Unknown error'
        };
      }
    }
    
    // Create snapshot
    expect(detectionResults).toMatchSnapshot();
    
    // Calculate success rate for assertion
    const totalFiles = Object.keys(detectionResults).length;
    const successfulDetections = Object.values(detectionResults).filter(r => r.magic !== null).length;
    const successRate = successfulDetections / totalFiles;
    
    // We expect at least 45% success rate since many text files don't have magic bytes
    expect(successRate).toBeGreaterThan(0.45);
  }, 60000); // 60 second timeout for this test

  test('should handle large files efficiently', async () => {
    const fixturePath = path.join(__dirname, '../../../../__fixtures__/kitchen-sink');
    const files = await glob('**/*', { cwd: fixturePath, nodir: true });
    
    // Filter for potentially large files
    const largeExtensions = ['pdf', 'zip', 'mp4', 'avi', 'mov', 'mkv', 'tar', 'gz'];
    const largeFiles = files.filter(file => {
      const ext = path.extname(file).toLowerCase().replace('.', '');
      return largeExtensions.includes(ext);
    });
    
    for (const file of largeFiles.slice(0, 5)) { // Test first 5 large files
      const filePath = path.join(fixturePath, file);
      const stream = createReadStream(filePath);
      
      const startTime = Date.now();
      const detectedType = await detector.detectFromStream(stream);
      const endTime = Date.now();
      
      // Should detect large files quickly (under 100ms)
      expect(endTime - startTime).toBeLessThan(100);
      
      // Detection should be fast for large files
    }
  });

  test('should detect specific file types correctly', async () => {
    const fixturePath = path.join(__dirname, '../../../../__fixtures__/kitchen-sink');
    
    // Test specific known file types
    const testCases = [
      { pattern: '*.png', expectedType: 'png', expectedMime: 'image/png' },
      { pattern: '*.jpg', expectedType: 'jpeg', expectedMime: 'image/jpeg' },
      { pattern: '*.jpeg', expectedType: 'jpeg', expectedMime: 'image/jpeg' },
      { pattern: '*.gif', expectedType: /^gif/, expectedMime: 'image/gif' },
      { pattern: '*.pdf', expectedType: 'pdf', expectedMime: 'application/pdf' },
      { pattern: '*.zip', expectedType: 'zip', expectedMime: 'application/zip' },
      { pattern: '*.mp3', expectedType: /^mp3/, expectedMime: 'audio/mpeg' },
      { pattern: '*.mp4', expectedType: 'mp4', expectedMime: 'video/mp4' }
    ];
    
    for (const testCase of testCases) {
      const files = await glob(testCase.pattern, { cwd: fixturePath, nodir: true });
      
      if (files.length > 0) {
        const file = files[0]; // Test first file of each type
        const filePath = path.join(fixturePath, file);
        const stream = createReadStream(filePath);
        
        const detectedType = await detector.detectFromStream(stream);
        
        expect(detectedType).toBeDefined();
        
        if (typeof testCase.expectedType === 'string') {
          expect(detectedType?.name).toBe(testCase.expectedType);
        } else {
          expect(detectedType?.name).toMatch(testCase.expectedType);
        }
        
        expect(detectedType?.mimeType).toBe(testCase.expectedMime);
        
        // Successfully detected the expected type
      }
    }
  });
});