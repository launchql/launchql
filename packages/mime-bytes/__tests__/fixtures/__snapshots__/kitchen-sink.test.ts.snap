// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Kitchen Sink Fixtures should detect all kitchen sink files correctly 1`] = `
{
  "LICENSE": {
    "contentType": "text/plain",
    "magic": {
      "charset": "utf-8",
      "confidence": 0.5,
      "name": "text",
      "type": "text/plain",
    },
  },
  "ai.ai": {
    "contentType": "application/pdf",
    "magic": {
      "charset": "ascii",
      "confidence": 1,
      "name": "pdf",
      "type": "application/pdf",
    },
  },
  "apng.png": {
    "contentType": "image/png",
    "magic": {
      "charset": "binary",
      "confidence": 1,
      "name": "png",
      "type": "image/png",
    },
  },
  "bmp.bmp": {
    "contentType": "image/bmp",
    "magic": {
      "charset": "binary",
      "confidence": 1,
      "name": "bmp",
      "type": "image/bmp",
    },
  },
  "build.js": {
    "contentType": "application/javascript",
    "magic": {
      "charset": "utf-8",
      "confidence": 0.8,
      "name": "text",
      "type": "text/plain",
    },
  },
  "build.js.map": {
    "contentType": "text/plain",
    "magic": {
      "charset": "utf-8",
      "confidence": 0.5,
      "name": "text",
      "type": "text/plain",
    },
  },
  "css.css": {
    "contentType": "text/css",
    "magic": {
      "charset": "utf-8",
      "confidence": 0.8,
      "name": "text",
      "type": "text/plain",
    },
  },
  "csv.csv": {
    "contentType": "text/csv",
    "magic": {
      "charset": "utf-8",
      "confidence": 0.8,
      "name": "text",
      "type": "text/plain",
    },
  },
  "docs.js": {
    "contentType": "application/javascript",
    "magic": {
      "charset": "utf-8",
      "confidence": 0.8,
      "name": "text",
      "type": "text/plain",
    },
  },
  "docs.jsx": {
    "contentType": "text/jsx",
    "magic": {
      "charset": "utf-8",
      "confidence": 0.8,
      "name": "text",
      "type": "text/plain",
    },
  },
  "docx.docx": {
    "contentType": "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
    "magic": {
      "charset": "binary",
      "confidence": 0.9,
      "name": "zip",
      "type": "application/zip",
    },
  },
  "dwg.dwg": {
    "contentType": "application/acad",
    "magic": {
      "charset": "binary",
      "confidence": 1,
      "name": "dwg",
      "type": "application/acad",
    },
  },
  "dxf.dxf": {
    "contentType": "image/vnd.dxf",
    "magic": {
      "charset": "utf-8",
      "confidence": 0.8,
      "name": "text",
      "type": "text/plain",
    },
  },
  "emf.emf": {
    "contentType": "image/emf",
    "magic": {
      "charset": "binary",
      "confidence": 1,
      "name": "emf",
      "type": "image/emf",
    },
  },
  "epub.epub": {
    "contentType": "application/epub+zip",
    "magic": {
      "charset": "binary",
      "confidence": 0.9,
      "name": "zip",
      "type": "application/zip",
    },
  },
  "font.otf": {
    "contentType": "font/otf",
    "magic": {
      "charset": "binary",
      "confidence": 1,
      "name": "otf",
      "type": "font/otf",
    },
  },
  "font.ttf": {
    "contentType": "font/ttf",
    "magic": {
      "charset": "binary",
      "confidence": 1,
      "name": "ttf",
      "type": "font/ttf",
    },
  },
  "font.woff": {
    "contentType": "font/woff2",
    "magic": {
      "charset": "binary",
      "confidence": 1,
      "name": "woff2",
      "type": "font/woff2",
    },
  },
  "font.woff2": {
    "contentType": "font/woff2",
    "magic": {
      "charset": "binary",
      "confidence": 1,
      "name": "woff2",
      "type": "font/woff2",
    },
  },
  "gif-w-alpha.gif": {
    "contentType": "image/gif",
    "magic": {
      "charset": "binary",
      "confidence": 1,
      "name": "gif89a",
      "type": "image/gif",
    },
  },
  "gif.gif": {
    "contentType": "image/gif",
    "magic": {
      "charset": "binary",
      "confidence": 1,
      "name": "gif89a",
      "type": "image/gif",
    },
  },
  "jpg.jpg": {
    "contentType": "image/jpeg",
    "magic": {
      "charset": "binary",
      "confidence": 1,
      "name": "jpeg",
      "type": "image/jpeg",
    },
  },
  "json.json": {
    "contentType": "application/json",
    "magic": {
      "charset": "utf-8",
      "confidence": 0.8,
      "name": "text",
      "type": "text/plain",
    },
  },
  "less.less": {
    "contentType": "text/x-less",
    "magic": {
      "charset": "utf-8",
      "confidence": 0.8,
      "name": "text",
      "type": "text/plain",
    },
  },
  "lock.lock": {
    "contentType": "text/plain",
    "magic": {
      "charset": "utf-8",
      "confidence": 0.5,
      "name": "text",
      "type": "text/plain",
    },
  },
  "md.md": {
    "contentType": "text/markdown",
    "magic": {
      "charset": "utf-8",
      "confidence": 0.8,
      "name": "text",
      "type": "text/plain",
    },
  },
  "mp4.mp4": {
    "contentType": "video/mp4",
    "magic": {
      "charset": "binary",
      "confidence": 1,
      "name": "mp4",
      "type": "video/mp4",
    },
  },
  "ods.ods": {
    "contentType": "application/vnd.oasis.opendocument.spreadsheet",
    "magic": {
      "charset": "binary",
      "confidence": 0.9,
      "name": "zip",
      "type": "application/zip",
    },
  },
  "odt.odt": {
    "contentType": "application/vnd.oasis.opendocument.text",
    "magic": {
      "charset": "binary",
      "confidence": 0.9,
      "name": "zip",
      "type": "application/zip",
    },
  },
  "pct.pct": {
    "contentType": "image/x-pict",
    "magic": {
      "charset": "binary",
      "confidence": 0.7,
      "name": "pct",
      "type": "image/x-pict",
    },
  },
  "pdf.pdf": {
    "contentType": "application/pdf",
    "magic": {
      "charset": "ascii",
      "confidence": 1,
      "name": "pdf",
      "type": "application/pdf",
    },
  },
  "png-w-alpha.png": {
    "contentType": "image/png",
    "magic": {
      "charset": "binary",
      "confidence": 1,
      "name": "png",
      "type": "image/png",
    },
  },
  "psd.psd": {
    "contentType": "image/vnd.adobe.photoshop",
    "magic": {
      "charset": "binary",
      "confidence": 1,
      "name": "psd",
      "type": "image/vnd.adobe.photoshop",
    },
  },
  "rtf.rtf": {
    "contentType": "application/rtf",
    "magic": {
      "charset": "utf-8",
      "confidence": 1,
      "name": "rtf",
      "type": "application/rtf",
    },
  },
  "scss.scss": {
    "contentType": "text/x-scss",
    "magic": {
      "charset": "utf-8",
      "confidence": 0.8,
      "name": "text",
      "type": "text/plain",
    },
  },
  "shellscript": {
    "contentType": "text/plain",
    "magic": {
      "charset": "utf-8",
      "confidence": 0.5,
      "name": "text",
      "type": "text/plain",
    },
  },
  "sql.sql": {
    "contentType": "text/plain",
    "magic": {
      "charset": "utf-8",
      "confidence": 1,
      "name": "utf8",
      "type": "text/plain",
    },
  },
  "svg-with-alpha-and-text.svg": {
    "contentType": "image/svg+xml",
    "magic": {
      "charset": "utf-8",
      "confidence": 0.8,
      "name": "text",
      "type": "text/plain",
    },
  },
  "svg.svg": {
    "contentType": "image/svg+xml",
    "magic": {
      "charset": "utf-8",
      "confidence": 0.8,
      "name": "text",
      "type": "text/plain",
    },
  },
  "svgo.yaml": {
    "contentType": "application/x-yaml",
    "magic": {
      "charset": "utf-8",
      "confidence": 0.8,
      "name": "text",
      "type": "text/plain",
    },
  },
  "swf.swf": {
    "contentType": "application/x-shockwave-flash",
    "magic": {
      "charset": "binary",
      "confidence": 1,
      "name": "swf-compressed",
      "type": "application/x-shockwave-flash",
    },
  },
  "tga.tga": {
    "contentType": "image/x-targa",
    "magic": {
      "charset": "binary",
      "confidence": 0.7,
      "name": "tga",
      "type": "image/x-targa",
    },
  },
  "tif.tif": {
    "contentType": "image/tiff",
    "magic": {
      "charset": "binary",
      "confidence": 1,
      "name": "tiff-be",
      "type": "image/tiff",
    },
  },
  "todo-0.1.0.tgz": {
    "contentType": "application/gzip",
    "magic": {
      "charset": "binary",
      "confidence": 1,
      "name": "gz",
      "type": "application/gzip",
    },
  },
  "tsv.tsv": {
    "contentType": "text/tab-separated-values",
    "magic": {
      "charset": "utf-8",
      "confidence": 0.8,
      "name": "text",
      "type": "text/plain",
    },
  },
  "txt.txt": {
    "contentType": "text/plain",
    "magic": {
      "charset": "utf-8",
      "confidence": 0.8,
      "name": "text",
      "type": "text/plain",
    },
  },
  "typescript.ts": {
    "contentType": "text/x-typescript",
    "magic": {
      "charset": "utf-8",
      "confidence": 0.8,
      "name": "text",
      "type": "text/plain",
    },
  },
  "typescript.tsx": {
    "contentType": "text/x-typescript",
    "magic": {
      "charset": "utf-8",
      "confidence": 0.8,
      "name": "text",
      "type": "text/plain",
    },
  },
  "wmf.wmf": {
    "contentType": null,
    "magic": null,
  },
  "xlsx.xlsx": {
    "contentType": "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
    "magic": {
      "charset": "binary",
      "confidence": 0.9,
      "name": "zip",
      "type": "application/zip",
    },
  },
  "zip.zip": {
    "contentType": "application/zip",
    "magic": {
      "charset": "binary",
      "confidence": 1,
      "name": "zip",
      "type": "application/zip",
    },
  },
}
`;
